### DISCUSSION TITLE
Launch at another port if the current is taken

### DISCUSSION THREAD

**Discussion by todortotev** [vercel/next.js#13487] [Ideas]
Excuse me if I'm missing something, but is it possible to have a command-line option if the port is taken? Seen it @ other apps so I thought it would come in handy.

For example, if port 3000 is taken, you get a message - Port XXXX is taken, do you want to use another port? Y/n

Currently, if the port is taken you get an error which also is okay, but I guess it would be even better if you can just click Y and get the app launched at a different port.

**Comment by Vadorequest** [comment]
Next.js automatically uses the next available port already.

If 3000 is taken, it tries 3001 and so on, until one free port is found.

**Reply by timneutkens** [reply]
This is incorrect. We fail when the port is already occupied.

**Reply by Vadorequest** [reply]
Indeed you're right! I got confused because I used `now dev` for quite a while and only recently switched to `next dev`, and I hadn't noticed this change of behaviour.

```bash
 ✔  yarn start
yarn run v1.22.4
$ next dev --port 8888
ready - started server on http://localhost:8888
info  - Loaded env from /Users/vadorequest/dev/next-right-now/.env.development.local
info  - Loaded env from /Users/vadorequest/dev/next-right-now/.env.development
Building Next with NODE_ENV="development" NEXT_PUBLIC_APP_STAGE="development" for NEXT_PUBLIC_CUSTOMER_REF="customer1"
warn  - You have enabled experimental feature(s).
warn  - Experimental features are not covered by semver, and may cause unexpected or broken application behavior. Use them at your own risk.

Port 8888 is already in use.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.

```

**Reply by todortotev** [reply]
☘️

**Reply by kkhcheng** [reply]
is there a way to go back to the old behavior? that seems more convenient in som cases

**Reply by eXamadeus** [reply]
Yeah, for those tuning in in 2023, Next.js now automatically grabs the next available port, for some reason.

Personally not a fan of automatically picking a port. I want to KNOW if there's an orphaned instance running. Why was this change made? I can't fathom a good reason for having a randomly assigned port. Sorry if I'm resurrecting a dead thread, but seems like Next.js was onboard with this for a while. I'd be interested to know the reasoning behind the change.

If the goal is to satisfy everyone, I feel like the automatic port selection can be left as the default, but turn it off when a port is specified. If I run `next dev -p 3000` and port 3000 is taken, I'd rather see an error than just randomly have Next.js popup at port 3001.

**Reply by dbrxnds** [reply]
Completely agree @eXamadeus. We just ran into this issue with production builds suddenly using another port than we assigned. How are we supposed to have it error on us?

**Reply by Kuroashi1995** [reply]
Does somebody knows if it at least exports the port where it's running so I can dynamically set a variable that I need?

**Reply by schoettler** [reply]
> Next.js automatically uses the next available port already.
> 
> If 3000 is taken, it tries 3001 and so on, until one free port is found.

Except when you pass a `-p` param for a custom port
